python/lib/sift_py/rule/service.py

```diff
@@ -0,0 +1,229 @@
+    CreateRuleRequest,
+    GetRuleRequest,
+    GetRuleResponse,
+    Rule,
+    RuleActionConfiguration,
+    RuleAssetConfiguration,
+    RuleConditionExpression,
+    UpdateActionRequest,
+    UpdateConditionRequest,
+    UpdateRuleRequest,
+)
+from sift.rules.v1.rules_pb2_grpc import RuleServiceStub
+
+from sift_py.grpc.transport import SiftChannel
+from sift_py.ingestion._internal.channel import channel_reference_from_fqn
+from sift_py.ingestion.channel import channel_fqn
+from sift_py.ingestion.config.yaml.load import load_rule_namespaces
+from sift_py.ingestion.rule.config import (
+    ExpressionChannelReference,
+    ExpressionChannelReferenceChannelConfig,
+    RuleActionCreateDataReviewAnnotation,
+    RuleActionCreatePhaseAnnotation,
+    RuleActionKind,
+    RuleConfig,
+)
+
+
+class RuleService:
+    _asset_service_stub: AssetServiceStub
+    _channel_service_stub: ChannelServiceStub
+    _rule_service_stub: RuleServiceStub
+
+    def __init__(self, channel: SiftChannel):
+        self._asset_service_stub = AssetServiceStub(channel)
+        self._channel_service_stub = ChannelServiceStub(channel)
+        self._rule_service_stub = RuleServiceStub(channel)
+
+    def load_rules_from_yaml(
+        self, paths: List[Path], sub_expressions: List[SubExpression]
+    ) -> List[RuleConfig]:
+        namespaced_rules = load_rule_namespaces(paths)
+
+        interpolation_map: Dict[str, Dict[str, Any]] = {}
+        for sub_expression in sub_expressions:
+            interpolation_map[sub_expression.fully_qualified_rule_name] = sub_expression.expressions
+
+        rule_configs = []
+        for namespace, rule_yamls in namespaced_rules.items():
+            for rule_yaml in rule_yamls:
+                channel_references: List[
+                    Union[ExpressionChannelReference, ExpressionChannelReferenceChannelConfig]
+                ] = []
+
+                for channel_ref in rule_yaml.get("channel_references", []):
+                    for ref, config in channel_ref.items():
+                        channel_references.append(
+                            {
+                                "channel_reference": ref,
+                                "channel_identifier": channel_fqn(
+                                    {
+                                        "channel_name": config.get("name", ""),
+                                        "component": config.get("component", ""),
+                                    }
+                                ),
+                            }
+                        )
+
+                rule_name = rule_yaml["name"]
+                rule_fqn = f"{namespace}.{rule_name}"
+                rule_subexpr = interpolation_map.get(rule_fqn, {})
+
+                rule_configs.append(
+                    RuleConfig(
+                        name=rule_yaml["name"],
+                        namespace=namespace,
+                        namespace_rules=namespaced_rules,
```

Comment:
Just FYI -- when you pass in a namespace & namespace_rules dict, `RuleConfig` will interpolate in the values from the namespaced_rules dict already: https://github.com/sift-stack/sift/blob/155df518efe4d4a23a22964a879e4a0b257c68ae/python/lib/sift_py/ingestion/rule/config.py#L81
So it potentially will overwrite whatever you passed in for description, expression, etc., although it should be all the same since you're also pulling them out of the same dict.

Reply:
oh that's actually so convenient! I'll include that in a cleanup PR. manually filling in all those fields was not fun.